### PR TITLE
WB-1568: Add theme utilities

### PR DESCRIPTION
**Pull request by jandrade**
## Summary:

- Adds `mergeTheme`: This utility is useful for merging themes together. It's a
bit like Object.assign, but it does a deep merge so that nested objects are
merged together and returned as a new object.

1. #1975
2. **WB-1568: Add theme utilities [CURRENT]**
3. #1978
4. #1979 
5. #1981
6. #1982
7. Create Storybook example for adding theming support to a custom component (WIP)

Issue: WB-1568

## Test plan:

Verify that the docs are correct and that the tests pass.

https://5e1bf4b385e3fb0020b7073c-tigkwckdoj.chromatic.com/?path=/docs/theming-mergetheme--docs
Screenshot 2023-08-10 at 6 02 37 PM

### CODE REVIEW

File: .changeset/tame-pots-brake.md [hunk 1]
```diff
@@ -0,0 +1,5 @@
+---
+"@khanacademy/wonder-blocks-theming": minor
+---
```
Comment (nishasy) [inline .changeset/tame-pots-brake.md:3]: Just curious - does this need two changesets?

File: __docs__/wonder-blocks-theming/merge-theme.mdx [hunk 2]
```diff
@@ -0,0 +1,75 @@
...
+defining another theme for a given component.
+
+```ts
+function mergeTheme<T>(sourceTheme: T, targetTheme: T): T;
+```
+
+<Banner
+    kind="info"
+    text="Note that the first argument will define the
+interface/contract that the second arg needs to follow."
```
Comment (nishasy) [inline __docs__/wonder-blocks-theming/merge-theme.mdx:23]: `layout` is a required prop on banner 😅

File: __docs__/wonder-blocks-theming/merge-theme.mdx [hunk 3]
```diff
@@ -0,0 +1,75 @@
...
+        radius: tokens.border.radius.large_6,
+    },
+});
+```
+
+Then the resulting themed button will be:
+
+<ButtonWithCustomTheme />
+
+And this is the resulting theme object:
```
Comment (nishasy) [inline __docs__/wonder-blocks-theming/merge-theme.mdx:59]: **Praise:** This is very helpful!

File: packages/wonder-blocks-theming/src/utils/__tests__/merge-theme.test.ts [hunk 4]
```diff
@@ -0,0 +1,87 @@
...
+            },
+        };
+
+        // Act
+        const result = mergeTheme(themeDefault, themeOverride);
+
+        // Assert
+        expect(result.color.blue).toEqual("#0000ff");
+    });
+});
```
Comment (nishasy) [inline packages/wonder-blocks-theming/src/utils/__tests__/merge-theme.test.ts:87]: In `merge-theme.mdx`, you have a note saying

> Note that the first argument will define the
interface/contract that the second arg needs to follow.

What happens if the second argument doesn't follow the interface properly? Is there a test for that?

File: __docs__/wonder-blocks-theming/create-theme-context.mdx [hunk 5]
```diff
@@ -0,0 +1,53 @@
...
+import {createThemeContext} from "@khanacademy/wonder-blocks-theming";
+
+import {ButtonWithTheme} from "./theme-examples";
+
+<Meta title="Theming / createThemeContext" />
+
+# `createThemeContext`
+
+`createThemeContext` is an utility that allow us to create a given theme that
+contains the component tokens associated to it.
```
Comment (nishasy) [inline __docs__/wonder-blocks-theming/create-theme-context.mdx:12]: Super tiny nit
```suggestion
`createThemeContext` is a utility that allows us to create a given theme that
contains the component tokens associated with it.
```

File: __docs__/wonder-blocks-theming/merge-theme.mdx [hunk 6]
```diff
@@ -0,0 +1,75 @@
...
+import Banner from "@khanacademy/wonder-blocks-banner";
+import {tokens, mergeTheme} from "@khanacademy/wonder-blocks-theming";
+
+import {ButtonWithCustomTheme} from "./theme-examples";
+
+<Meta title="Theming / mergeTheme" />
+
+# `mergeTheme`
+
+`mergeTheme` is an utility that allow us to create a new copy of the target
```
Comment (nishasy) [inline __docs__/wonder-blocks-theming/merge-theme.mdx:12]: Super tiny nit
```suggestion

`mergeTheme` is a utility that allows us to create a new copy of the target
```

File: packages/wonder-blocks-theming/src/utils/create-theme-context.ts [hunk 7]
```diff
@@ -0,0 +1,12 @@
+import * as React from "react";
+
+/**
+ * An utility that allow us to create a given theme that contains the component
+ * tokens associated to it.
```
Comment (nishasy) [inline packages/wonder-blocks-theming/src/utils/create-theme-context.ts:5]: Super tiny nit
```suggestion
 * A utility that allows us to create a given theme that contains the component
 * tokens associated with it.
```

File: packages/wonder-blocks-theming/src/utils/merge-theme.ts [hunk 8]
```diff
@@ -0,0 +1,33 @@
+type RecursivePartial<T> = {
+    [P in keyof T]?: RecursivePartial<T[P]> | string | number | boolean;
+};
+
+/**
+ * Allow us to create a new copy of the target theme by overriding some of its
```
Comment (nishasy) [inline packages/wonder-blocks-theming/src/utils/merge-theme.ts:6]: teeniest tiniest nit
```suggestion
 * Allows us to create a new copy of the target theme by overriding some of its
```

**Review by nishasy — APPROVED** [review]
🎉 I love it! Looks great! A couple of reeeaaaalllly small nits - feel free to ignore them.

File: __docs__/wonder-blocks-theming/create-theme-context.mdx [hunk 9]
```diff
@@ -0,0 +1,53 @@
...
+
+const ButtonThemeContext = createThemeContext(buttonDefaultTheme);
+
+const ThemedButton = () => {
+    const theme = React.useContext(ButtonThemeContext);
+    return (
+        <button
+            style={{
+                background: theme.color.bg.primary,
+                borderRadius: theme.border.radius,
```
Comment (nishasy) [inline __docs__/wonder-blocks-theming/create-theme-context.mdx:42]: Could you add an explanation to these docs about why using `createThemeContext` is better than just doing
```
background: buttonDefaultTheme.color.bg.primary,
borderRadius: buttonDefaultTheme.border.radius,
```

File: packages/wonder-blocks-theming/src/utils/__tests__/merge-theme.test.ts (hunk 4 above)
Comment (jandrade) [reply packages/wonder-blocks-theming/src/utils/__tests__/merge-theme.test.ts:87]: TS will throw an error :)

e.g.
Screenshot 2023-08-14 at 11 07 37 AM

File: .changeset/tame-pots-brake.md (hunk 1 above)
Comment (jandrade) [reply .changeset/tame-pots-brake.md:3]: 🤔 probably not, I was planning to only add one thing, but then decided to send to commits into the same PR and that caused pushing both files. The good news is that changesets should be able to cover that.

**Comment by jandrade** [comment]
The parent pull-request ([#1975](https://github.com/Khan/wonder-blocks/pull/1975)) has been merged into `main`, but this branch (`theming-utils`) now has conflicts with the new base branch. These conflicts must be resolved before checks can complete on this pull-request.

File: packages/wonder-blocks-theming/src/utils/__tests__/merge-theme.test.ts (hunk 4 above)
Comment (nishasy) [reply packages/wonder-blocks-theming/src/utils/__tests__/merge-theme.test.ts:87]: Oh perfect!

File: __docs__/wonder-blocks-theming/create-theme-context.mdx [hunk 10]
```diff
@@ -8,9 +8,13 @@ import {ButtonWithTheme} from "./theme-examples";
 
 # `createThemeContext`
 
-`createThemeContext` is an utility that allow us to create a given theme that
+`createThemeContext` is a utility that allows us to create a given theme that
 contains the component tokens associated to it.
 
+This is useful when we want to create a component that can be themed, and we
+want to provide a default theme for it. Also, this allows us to extend the
+component with more themes and swtich between them using the `ThemeProvider`.
```
Comment (nishasy) [inline __docs__/wonder-blocks-theming/create-theme-context.mdx:16]:
```suggestion
component with more themes and switch between them using the `ThemeProvider`.
```

**Pull request merged**